File: scripts/jobs/check_extraction_success.js
#!/usr/bin/env node
/**
 * Task: Check Extraction Success
 * Purpose: Verify that quarterly-result-extractor (or guidance-document-extractor)
 * persisted records to the DB successfully. Used as a gate before downstream analysis
 * skills (quarterly-result-analysis, forward-guidance-extractor).
 *
 * Usage:
 *   node check_extraction_success.js --collection quarterly-result-documents --date 2026-08-11
 *   node check_extraction_success.js --collection guidance-documents --date 2026-08-11
 *
 * Exit codes:
 *   0 = Success (records found, analysis task can proceed)
 *   1 = Failure (no records, analysis task will NOT run)
 */

'use strict';

const path = require('path');

/**
 * Parse CLI args to extract named parameters.
 */
function argValue(argv, key, defaultVal = null) {
  const idx = argv.indexOf(key);
  return idx >= 0 && idx + 1 < argv.length ? argv[idx + 1] : defaultVal;
}

/**
 * Check if records were persisted to the DB for the given date.
 *
 * NOTE: 'quarterly-result-documents' and 'guidance-documents' are DTO `type`
 * values, not db.js collection names — both are persisted via
 * `db.saveReport()` into the single `reports` collection (see
 * docs/DATA_RULES.md §2, "Analysis/report DTO" row; db.js's
 * SINGLE_FILE_COLLECTIONS list does not include either of these names, so
 * calling `db.find(collectionName, ...)` directly throws "Unknown
 * collection"). Always query the `reports` collection and filter on `type`.
 *
 * @param {string} collectionName - the DTO `type`, e.g. 'quarterly-result-documents', 'guidance-documents'
 * @param {string} dateStr - Date string in YYYY-MM-DD format
 * @returns {Promise<{status: 'success'|'failure'|'error', count: number, message: string, date: string, records?: Array}>}
 */
async function checkDbRecords(collectionName, dateStr) {
  try {
    // Dynamically load db.js from the repo's jobs-runtime
    // This script runs in the Cowork environment where the repo is mounted
    const db = require('../../packages/jobs-runtime/lib/db');

    // Query for records created on the given date.
    // db.js's find() does exact-equality matching on `creator` only (no $in
    // support), so fetch by type+date and post-filter for either extraction
    // skill's creator here.
    //
    // IMPORTANT: db.find() returns thin index entries only (id/type/date/
    // companyId/creator/summary/body — see db.js's `find()` doc comment).
    // It does NOT include content fields like `excerpts` or `found` — those
    // live in the full body file that `body` points to and require
    // `db.readReport(id)` to load. A prior version of this script counted
    // any matching index entry as "success," which silently passed even
    // when guidance-document-extractor had only fetched documents but not
    // yet run its relevance-filter pass (excerptsPending: true, excerpts: []
    // on every record) — the exact "fetched but not filtered" state this
    // gate exists to catch. Always read the full body before judging
    // completeness.
    const byTypeAndDate = db.find('reports', { type: collectionName, date: dateStr });
    const candidates = byTypeAndDate.filter((r) =>
      ['quarterly-result-extractor', 'guidance-document-extractor'].includes(r.creator)
    );

    if (!candidates || candidates.length === 0) {
      return {
        status: 'failure',
        count: 0,
        message: `No ${collectionName} records persisted for ${dateStr}`,
        date: dateStr,
      };
    }

    // Load full bodies and require the extraction to have actually finished
    // (excerptsPending === false). A record can legitimately end up with
    // excerpts: [] (Stage 2 ran and genuinely found nothing — see
    // forward-guidance-extractor's SKILL.md case 3), but excerptsPending
    // still lets us distinguish that from "Stage 1 fetched documents and
    // stopped before the relevance-filter pass ran," which is a failure
    // this gate must catch, not silently pass through.
    const fullRecords = candidates.map((r) => db.readReport(r.id));
    const incomplete = fullRecords.filter((r) => r.excerptsPending === true);
    const complete = fullRecords.filter((r) => r.excerptsPending !== true);

    if (complete.length === 0) {
      return {
        status: 'failure',
        count: 0,
        message: `Found ${candidates.length} ${collectionName} records for ${dateStr}, but all ${incomplete.length} still have excerptsPending: true (relevance-filter pass never completed)`,
        date: dateStr,
        incompleteIds: incomplete.map((r) => r.id).slice(0, 10),
      };
    }

    const result = {
      status: 'success',
      count: complete.length,
      message: `Found ${complete.length} completed ${collectionName} records for ${dateStr}`,
      date: dateStr,
      records: complete.slice(0, 3), // Return first 3 full records as sample
    };
    if (incomplete.length > 0) {
      result.message += ` (${incomplete.length} of ${candidates.length} still have excerptsPending: true and were excluded)`;
      result.incompleteIds = incomplete.map((r) => r.id);
    }
    return result;
  } catch (error) {
    // If db module can't be loaded, try an alternative approach using fs directly
    try {
      const fs = require('fs');

      // Try to read the reports.json file directly
      const reportPath = path.join(
        __dirname,
        '../../data/reports.json'
      );

      if (!fs.existsSync(reportPath)) {
        return {
          status: 'error',
          count: 0,
          error: `Reports database not found at ${reportPath}`,
          date: dateStr,
        };
      }

      const reportsContent = fs.readFileSync(reportPath, 'utf8');
      const reports = JSON.parse(reportsContent);

      // data/reports.json is a flat object keyed by record id (index
      // entries, same thin shape as db.find() — see the note above), NOT
      // { records: [...] }. Object.values() is the correct read.
      const matching = Object.values(reports).filter(
        (r) =>
          r.type === collectionName &&
          r.date === dateStr &&
          (r.creator === 'quarterly-result-extractor' ||
            r.creator === 'guidance-document-extractor')
      );

      if (!matching.length) {
        return {
          status: 'failure',
          count: 0,
          message: `No ${collectionName} records persisted for ${dateStr}`,
          date: dateStr,
        };
      }

      // Same completeness requirement as the primary path: an index entry
      // alone doesn't tell us whether Stage 2's relevance-filter finished,
      // so load each full body (reports/<id>.json, per the `body` field)
      // and check excerptsPending there.
      const dataDir = path.join(__dirname, '../../data');
      const fullMatching = matching.map((r) => {
        try {
          return JSON.parse(fs.readFileSync(path.join(dataDir, r.body), 'utf8'));
        } catch {
          return r; // body file missing/unreadable — fall back to the index entry as-is
        }
      });
      const incomplete = fullMatching.filter((r) => r.excerptsPending === true);
      const complete = fullMatching.filter((r) => r.excerptsPending !== true);

      if (complete.length === 0) {
        return {
          status: 'failure',
          count: 0,
          message: `Found ${matching.length} ${collectionName} records for ${dateStr}, but all ${incomplete.length} still have excerptsPending: true (relevance-filter pass never completed)`,
          date: dateStr,
          incompleteIds: incomplete.map((r) => r.id).slice(0, 10),
        };
      }

      const fallbackResult = {
        status: 'success',
        count: complete.length,
        message: `Found ${complete.length} completed ${collectionName} records for ${dateStr}`,
        date: dateStr,
        records: complete.slice(0, 3),
      };
      if (incomplete.length > 0) {
        fallbackResult.message += ` (${incomplete.length} of ${matching.length} still have excerptsPending: true and were excluded)`;
        fallbackResult.incompleteIds = incomplete.map((r) => r.id);
      }
      return fallbackResult;
    } catch (fallbackError) {
      return {
        status: 'error',
        error: `Could not load db module or read reports.json: ${error.message}`,
        fallbackError: fallbackError.message,
        date: dateStr,
      };
    }
  }
}

/**
 * Main entry point.
 */
async function main() {
  const collectionName = argValue(process.argv, '--collection');
  const dateStr = argValue(process.argv, '--date');

  if (!collectionName || !dateStr) {
    console.error(
      JSON.stringify(
        {
          status: 'error',
          error: 'Usage: node check_extraction_success.js --collection <name> --date <YYYY-MM-DD>',
        },
        null,
        2
      )
    );
    process.exit(1);
  }

  const result = await checkDbRecords(collectionName, dateStr);
  console.log(JSON.stringify(result, null, 2));

  // Exit with failure if status is not 'success'
  if (result.status !== 'success') {
    process.exit(1);
  }
}

if (require.main === module) {
  main();
}

module.exports = { checkDbRecords };

File: scripts/jobs/check_extraction_success.py
#!/usr/bin/env python3
"""
Task: Check Extraction Success
Purpose: Verify that quarterly-result-extractor (or guidance-document-extractor)
persisted records to the DB successfully. Used as a gate before downstream analysis
skills (quarterly-result-analysis, forward-guidance-extractor).

Usage:
  python check_extraction_success.py --collection quarterly-result-documents --date 2026-08-11
  python check_extraction_success.py --collection guidance-documents --date 2026-08-11
"""

import sys
import json
import os
from datetime import datetime

# Add repo to path for db access
sys.path.insert(0, "/Users/darshanpatel/code/stockmarket/packages/jobs-runtime")

def check_db_records(collection_name, date_str):
    """
    Check if records were persisted to the DB for the given date.

    Args:
        collection_name: e.g. 'quarterly-result-documents', 'guidance-documents'
        date_str: Date string in YYYY-MM-DD format

    Returns:
        {status: 'success'|'failure', count: N, message: '...'}
    """
    try:
        # Query reports.json directly (simpler than importing Node db module)
        import json
        import os

        # Find repo root by looking for data/reports.json
        script_dir = os.path.dirname(os.path.abspath(__file__))
        repo_root = os.path.dirname(os.path.dirname(script_dir))
        reports_path = os.path.join(repo_root, 'data', 'reports.json')

        with open(reports_path, 'r') as f:
            reports = json.load(f)

        # data/reports.json is a flat object keyed by record id containing
        # thin INDEX entries (id/type/date/companyId/creator/summary/body) —
        # not full record bodies. Content fields like `excerpts` and
        # `excerptsPending` live in the full body file that `body` points to
        # (data/reports/<id>.json) and must be read separately. An earlier
        # version of this script counted any matching index entry as
        # "success," which silently passed even when
        # guidance-document-extractor had only fetched documents but never
        # run its relevance-filter pass (excerptsPending: true, excerpts: []
        # on every record) — exactly the "fetched but not filtered" state
        # this gate exists to catch.
        matching_records = [
            r for r in reports.values()
            if r.get('type') == collection_name
            and r.get('date') == date_str
            and r.get('creator') in [
                'quarterly-result-extractor',
                'guidance-document-extractor'
            ]
        ]

        if not matching_records:
            return {
                "status": "failure",
                "count": 0,
                "message": f"No {collection_name} records persisted for {date_str}",
                "date": date_str
            }

        # Load full bodies and require the extraction to have actually
        # finished (excerptsPending is not True). excerpts: [] alone can be
        # a legitimate "attempted, genuinely nothing found" outcome (see
        # forward-guidance-extractor's SKILL.md case 3) — excerptsPending is
        # what distinguishes that from an unfinished Stage 1 run.
        full_records = []
        for r in matching_records:
            body_path = os.path.join(repo_root, 'data', r.get('body', ''))
            try:
                with open(body_path, 'r') as bf:
                    full_records.append(json.load(bf))
            except (OSError, IOError, ValueError):
                full_records.append(r)  # body file missing/unreadable — fall back to index entry

        incomplete = [r for r in full_records if r.get('excerptsPending') is True]
        complete = [r for r in full_records if r.get('excerptsPending') is not True]

        if not complete:
            return {
                "status": "failure",
                "count": 0,
                "message": f"Found {len(matching_records)} {collection_name} records for {date_str}, "
                           f"but all {len(incomplete)} still have excerptsPending: true "
                           f"(relevance-filter pass never completed)",
                "date": date_str,
                "incompleteIds": [r.get('id') for r in incomplete[:10]]
            }

        result = {
            "status": "success",
            "count": len(complete),
            "message": f"Found {len(complete)} completed {collection_name} records for {date_str}",
            "date": date_str,
            "firstRecord": complete[0] if complete else None
        }
        if incomplete:
            result["message"] += f" ({len(incomplete)} of {len(matching_records)} still have excerptsPending: true and were excluded)"
            result["incompleteIds"] = [r.get('id') for r in incomplete]
        return result

    except Exception as e:
        return {
            "status": "error",
            "error": str(e),
            "date": date_str
        }

def main():
    if len(sys.argv) < 5:
        print(json.dumps({
            "status": "error",
            "error": "Usage: python check_extraction_success.py --collection <name> --date <YYYY-MM-DD>"
        }), file=sys.stderr)
        sys.exit(1)

    collection = None
    date_str = None

    for i in range(1, len(sys.argv), 2):
        if sys.argv[i] == "--collection":
            collection = sys.argv[i + 1]
        elif sys.argv[i] == "--date":
            date_str = sys.argv[i + 1]

    if not collection or not date_str:
        print(json.dumps({
            "status": "error",
            "error": "Missing --collection or --date"
        }), file=sys.stderr)
        sys.exit(1)

    result = check_db_records(collection, date_str)
    print(json.dumps(result, indent=2))

    # Exit with failure if status is not 'success'
    if result["status"] != "success":
        sys.exit(1)

if __name__ == "__main__":
    main()
